skada/utils.py

```diff
@@ -247,3 +251,223 @@ def source_target_split(
+            arrays[i] = np.array([])
+
+        if arrays[i+1] is None or arrays[i+1].shape[0] == 0:
+            arrays[i+1] = np.array([])
+
+        # Check no duo is empty
+        if (np.size(arrays[i]) == 0 and np.size(arrays[i+1]) == 0):
+            raise ValueError("Only one array can be None or empty in each duo")
+
+        # Check consistent dim of arrays
+        if (np.size(arrays[i]) != 0 and np.size(arrays[i+1]) != 0):
+            if arrays[i].shape[1:] != arrays[i+1].shape[1:]:
+                raise ValueError(
+                    "Inconsistent number of features in source-target arrays"
+                )
+
+    # ################ Check sample_domain #################
+    # If sample_domain is None, we need to infer it from the target array
+    if sample_domain is None or sample_domain.shape[0] == 0:
+        # We need to infer the domain from the target array
+        warnings.warn(
+            "sample_domain is None or empty, it will be inferred from the arrays"
+        )
+
+        # By assuming that the first array is the source and the second the target
+        index_something_source = 0
+        index_something_target = 1
+        sample_domain = np.concatenate((
+            _DEFAULT_SOURCE_DOMAIN_LABEL*np.ones(
+                arrays[index_something_source].shape[0]
+            ),
+            _DEFAULT_TARGET_DOMAIN_LABEL*np.ones(
+                arrays[index_something_target].shape[0]
+            )
+        ))
+
+    # To test afterward if the number of samples in source-target arrays
+    # and the number infered in the sample_domain are consistent
+    source_indices = extract_source_indices(sample_domain)
+    source_samples = np.count_nonzero(source_indices)
+    target_samples = np.count_nonzero(~source_indices)
+
+    # ################ Merge arrays #################
+    merges = []  # List of merged arrays
+
+    for i in range(0, len(arrays), 2):
+
+        # If one of the array is empty, we need to infer its values
+        index_is_empty = None  # Index of the array that was None
+        if np.size(arrays[i]) == 0:
+            index_is_empty = i
+
+        if np.size(arrays[i+1]) == 0:
+            index_is_empty = i+1
+
+        if index_is_empty is not None:
+            # We need to infer the value of the empty array in the duo
+            warnings.warn(
+                "One of the arrays in a duo is empty, it will be inferred"
+            )
+
+            duo_index = i+1 if index_is_empty == i else i
+
+            y_type = type_of_target(arrays[duo_index])
+            if y_type == 'binary' or y_type == 'multiclass':
+                arrays[index_is_empty] = (
+                    _DEFAULT_MASKED_TARGET_CLASSIFICATION_LABEL *
+                    np.ones(
+                        (sample_domain.shape[0] - arrays[duo_index].shape[0],) +
+                        arrays[duo_index].shape[1:]
+                    )
+                )
+            else:
```

Comment:
The only difference between it and else branches is the value of the label. Thus the code could be simplified.